ResearchKit/Common/ORKTaskViewController.m

```diff
@@ -893,9 +893,18 @@ - (void)showViewController:(ORKStepViewController *)viewController goForward:(BO
         }
     }
     
-    if (step.identifier && ![_managedStepIdentifiers.lastObject isEqualToString:step.identifier]) {
-        [_managedStepIdentifiers addObject:step.identifier];
+    NSString *stepIdentifier = step.identifier;
+    if (stepIdentifier) {
+        NSUInteger indexOfStep = [_managedStepIdentifiers indexOfObject:stepIdentifier];
+        if (indexOfStep == NSNotFound) {
+            [_managedStepIdentifiers addObject:stepIdentifier];
+        } else {
+            // If the step being shown has been visited before, remove all the managed step identifiers visited after it
+            NSUInteger numberManagedOfStepIdentifiers = _managedStepIdentifiers.count;
+            [_managedStepIdentifiers removeObjectsInRange:NSMakeRange(indexOfStep + 1, numberManagedOfStepIdentifiers - (indexOfStep + 1))];
```

Comment:
Ok, I have reverted the behaviour to the previous push/pop stack model, so all the navigation stack (including loops) is preserved. As discussed, the results for revisited steps appear duplicated on the final task result, but all of them point to the same result instance (which correspond to the user's latest answer).

I have also extended the developer warning mentioning this (produced from `ORKNavigableOrderedTask`), and added some *headerdoc* paragraphs explaining this.